pkg/internal/templates.go

```diff
@@ -0,0 +1,213 @@
+
+func (w *wrapped{{.Name}}) checkCluster(ctx context.Context) (context.Context, error) {
+	ctxCluster, ok := kcp.ClusterFromContext(ctx)
+	if !ok {
+		return kcp.WithCluster(ctx, w.cluster), nil
+	} else if ctxCluster != w.cluster {
+		return ctx, fmt.Errorf("cluster mismatch: context=%q, client=%q", ctxCluster, w.cluster)
+	}
+	return ctx, nil
+}
+`
+
+const wrapperMethodsTempl = `
+func (w *wrapped{{.PkgNameUpperFirst}}{{.VersionUpperFirst}}) {{.Name}}s() {{.PkgName}}{{.Version}}.{{.Name}}Interface {
+	return &wrapped{{.Name}}{
+		cluster:  w.cluster,
+		delegate: w.delegate.{{.Name}}s(),
+	}
+}
+
+type wrapped{{.Name}} struct {
+	cluster  string
+	delegate {{.PkgName}}.{{.Name}}Interface
+}
+
+func (w *wrapped{{.Name}}) Create(ctx context.Context, {{.NameLowerFirst}} *{{.PkgName}}api{{.Version}}.{{.Name}}, opts metav1.CreateOptions) (*{{.PkgName}}api{{.Version}}.{{.Name}}, error) {
+	ctx, err := w.checkCluster(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return w.delegate.Create(ctx, {{.NameLowerFirst}}}, opts)
+}
+
+func (w *wrapped{{.Name}}) Update(ctx context.Context, {{.NameLowerFirst}}} *{{.PkgName}}api{{.Version}}.{{.Name}}, opts metav1.UpdateOptions) (*{{.PkgName}}api{{.Version}}.{{.Name}}, error) {
+	ctx, err := w.checkCluster(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return w.delegate.Update(ctx, {{.NameLowerFirst}}}, opts)
+}
+
+func (w *wrapped{{.Name}}) Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error {
+	ctx, err := w.checkCluster(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return w.delegate.Delete(ctx, name, opts)
+}
+
+func (w *wrapped{{.Name}}) DeleteCollection(ctx context.Context, opts metav1.DeleteOptions, listOpts metav1.ListOptions) error {
+	ctx, err := w.checkCluster(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return w.delegate.Delete(ctx, opts, listOpts)
+}
+
+func (w *wrapped{{.Name}}) Get(ctx context.Context, name string, opts metav1.GetOptions) (*{{.PkgName}}api{{.Version}}.{{.Name}}, error) {
+	ctx, err := w.checkCluster(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return w.delegate.Get(ctx, opts, listOpts)
+}
+
+func (w *wrapped{{.Name}}) List(ctx context.Context, opts metav1.ListOptions) (*{{.PkgName}}api{{.Version}}.{{.Name}}List, error) {
+	ctx, err := w.checkCluster(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return w.delegate.List(ctx, opts)
+}
+
+func (w *wrapped{{.Name}}) Watch(ctx context.Context, opts metav1.ListOptions) (watch.Interface, error) {
+	ctx, err := w.checkCluster(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return w.delegate.Watch(ctx, opts)
+}
+
+func (w *wrapped{{.Name}}) Patch(ctx context.Context, name string, pt apiTypes.PatchapiType, data []byte, opts metav1.PatchOptions, subresources ...string) (result *{{.PkgName}}api{{.Version}}.{{.Name}}, err error) {
+	ctx, err := w.checkCluster(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return w.delegate.Patch(ctx, name, pt, data, opts, subresources)
+}
+`
+
+// Do we make this configurable?
```

Comment:
The other generators typically use a `--go-header-file` flag. We will want to do something similar.

Reply:
Added option to accept header file.